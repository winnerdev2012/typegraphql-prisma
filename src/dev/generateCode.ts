import { DMMF } from "@prisma/photon/dist/runtime/dmmf-types";
import { Project } from "ts-morph";

import generateEnumsFromDef from "../../src/generator/enums";
import generateObjectTypeClassesFromModel from "../../src/generator/object-type-classes";
import generateImports from "../../src/generator/imports";
import generateRelationsResolverClassesFromModel from "../../src/generator/relations-resolver-classes";

export async function generateCode(
  dataModelJSONPath: string,
  outputTSFilePath: string,
) {
  console.log("Loading datamodel...");
  const dmmf = {
    datamodel: require(dataModelJSONPath) as DMMF.Datamodel,
  };

  const project = new Project();
  const sourceFile = project.createSourceFile(outputTSFilePath, undefined, {
    overwrite: true,
  });

  console.log("Generating imports...");
  await generateImports(sourceFile);

  console.log("Generating enums...");
  await Promise.all(
    dmmf.datamodel.enums.map(enumDef =>
      generateEnumsFromDef(sourceFile, enumDef),
    ),
  );

  console.log("Generating models...");
  await Promise.all(
    dmmf.datamodel.models.map(model =>
      generateObjectTypeClassesFromModel(sourceFile, model),
    ),
  );

  console.log("Generating relation resolvers...");
  await Promise.all(
    dmmf.datamodel.models.map(model =>
      generateRelationsResolverClassesFromModel(sourceFile, model),
    ),
  );

  console.log("Formating generated code...");
  // sourceFile.fixUnusedIdentifiers();
  sourceFile.formatText({ indentSize: 2 });

  console.log("Saving generated code...");
  await sourceFile.save();
}
