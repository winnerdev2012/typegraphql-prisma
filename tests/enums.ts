import { promises as fs } from "fs";

import generateCode from "../src/generator/generate-code";
import getPhotonDmmfFromPrismaSchema from "./helpers/dmmf";
import getBaseDirPath from "./helpers/base-dir";

describe("enums", () => {
  let baseDirPath: string;

  beforeEach(() => {
    baseDirPath = getBaseDirPath("enums");
  });

  afterEach(async () => {
    await fs.rmdir(baseDirPath, { recursive: true });
    await new Promise(r => setTimeout(r, 100));
  });

  it("should properly generate code for normal enum", async () => {
    const schema = /* prisma */ `
      enum Color {
        RED
        GREEN
        BLUE
      }
    `;

    await generateCode(
      await getPhotonDmmfFromPrismaSchema(schema),
      baseDirPath,
    );
    const colorEnumTSFile = await fs.readFile(baseDirPath + "/enums/color.ts", {
      encoding: "utf-8",
    });

    expect(colorEnumTSFile).toMatchSnapshot();
  });

  it("should properly generate code for enum with docs", async () => {
    const schema = /* prisma */ `
      /// Role enum doc
      enum Role {
        // User member comment
        USER
        ADMIN
      }
    `;

    await generateCode(
      await getPhotonDmmfFromPrismaSchema(schema),
      baseDirPath,
    );
    const roleEnumTSFile = await fs.readFile(baseDirPath + "/enums/role.ts", {
      encoding: "utf-8",
    });

    expect(roleEnumTSFile).toMatchSnapshot();
  });
});
