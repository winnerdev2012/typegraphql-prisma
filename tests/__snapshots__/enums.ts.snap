// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enums should properly generate code for enum with docs 1`] = `
"import { Arg, Args, ArgsType, Ctx, Field, FieldResolver, Float, ID, InputType, Int, Mutation, ObjectType, Query, Resolver, Root, registerEnumType } from \\"type-graphql\\";

/** Role enum doc */
export enum Role {
  USER = \\"USER\\",
  ADMIN = \\"ADMIN\\"
}
registerEnumType(Role, {
  name: \\"Role\\",
  description: \\"Role enum doc\\",
});
"
`;

exports[`enums should properly generate code for normal enum 1`] = `
"import { Arg, Args, ArgsType, Ctx, Field, FieldResolver, Float, ID, InputType, Int, Mutation, ObjectType, Query, Resolver, Root, registerEnumType } from \\"type-graphql\\";

export enum Color {
  Red = \\"Red\\",
  Green = \\"Green\\",
  Blue = \\"Blue\\"
}
registerEnumType(Color, {
  name: \\"Color\\",
  description: undefined,
});
"
`;
